### PR TITLE
CO: Restore from deleted order state

### PR DESCRIPTION
**Pull request by firstred**
## Summary

| Questions | Answers |
| --- | --- |
| Branch? | `1.6.1.x` |
| Description? | Order states can be deleted. If that happens and the merchant tries to change the order state, then it is no longer possible to change the order state of an order in the BO. |
| Type? | bug fix |
| Category? | CO |
| BC breaks? | no |
| Deprecations? | no |
| Fixed ticket? |  |
| How to test? | \- Create order, delete current order status, change order status from BO - should be fine - Remove `not null` restriction on `current_state` column in `ps_orders` as this situation can occur somehow (maybe we could fix this with an upgrade script?), make `current_state` `null` and update the Order state of the order. This should be fine as well |
## Stacktrace

``` php
PHP Notice·Trying to get property of non-object
-------------------
/home/anonymous/domains/example.com/public_html/controllers/admin/AdminOrdersController.php:521AdminOrdersControllerCore::postProcess   
                     $this->errors[] = Tools::displayError('The new order status is invalid.');
                } else {
                    $current_order_state = $order->getCurrentOrderState();
                    if ($current_order_state->id != $order_state->id) {
                        // Create new OrderHistory
                        $history = new OrderHistory();
                        $history->id_order = $order->id;
-------------------
/home/anonymous/domains/example.com/public_html/classes/controller/Controller.php:178ControllerCore::run    
             }
            // postProcess handles ajaxProcess
            $this->postProcess();
            if (!empty($this->redirect_after)) {
                $this->redirect();
-------------------
/home/anonymous/domains/example.com/public_html/classes/Dispatcher.php:367DispatcherCore::dispatch  
             }
            // Running controller
            $controller->run();
        } catch (PrestaShopException $e) {
            $e->displayMessage();
        }
-------------------
/home/anonymous/domains/example.com/public_html/roxcwhiki8trxuvj/index.php:63[main] 
 if (!isset($_REQUEST['controller']) && isset($_REQUEST['tab'])) {
    $_REQUEST['controller'] = strtolower($_REQUEST['tab']);
}
// Prepare and trigger admin dispatcher
Dispatcher::getInstance()->dispatch();
```

### CODE REVIEW

**Comment by vincentbz** [comment]
Hi @firstred ,

Thanks :)
That's weird, I already can change the status of an order after I deleted the status, on the 1.6.1.6, without your fix.
Did i do something wrong ? 
1. i create a status
2. i change an order to that status
3. i delete the status
4. i change the order status

Maybe it's related to a status where there is an invoice or something like this ?

delete status

**Comment by firstred** [comment]
It's just a PHP notice, so I think that PHP is able to recover in most cases, but PrestaShop should handle it nicely without causing log messages.

Do you see any errors (notices) in your server log?

**Comment by vincentbz** [comment]
Yep, I have this:
[15-Jun-2016 18:23:35 Europe/Paris] PHP Fatal error:  Uncaught  --> Smarty: unable to write file C:\wamp\www\prestashop1616-1/cache/smarty/compile\82\bc\0b\82bc0b85667a9ee4b041811a221f3f7eff577b97.file.notification_bt.tpl.php <-- 
  thrown in C:\wamp\www\prestashop1616-1\tools\smarty\sysplugins\smarty_internal_write_file.php on line 74

**Comment by vincentbz** [comment]
If I apply your PR : 
1. I create a status
2. i change an order to that status
3. i delete the status

Then, the status has disappeard in the order and I have something different, I'm set on the first of the list

image

**Comment by firstred** [comment]
**tl;dr - `$current_order_state !== null` should be used instead**

Hmmz, just discovered that `Validate::isLoadedObject` may not be used.

I examined the production environment and the following happens:
The current state id in the DB is `null`, despite the fact that the DB column in the table `ps_orders` has a `NOT NULL` restriction, which means that `$order->getCurrentOrderState()` returns `null` in that case.

When deleting an `OrderState`, and updating an order with that `OrderState`, the check `Validate::isLoadedObject` will fail, despite that fact that it used to be a valid state in the past, so this may not be used to check the `OrderState`.

It is better to use the hard check `$current_order_state !== null` because this is the only check that can compensate for the wrong value in the database table.

So to summarize:
It should not be possible to have a `null` value in the DB column `order_state` in the table `ps_orders`, but somehow does occur (broken upgrade scripts in the past maybe), but seen the fact that `Order::getCurrentOrderState()` can return `null`, I think that it wouldn't hurt to add another check.

I hope that haven't fallen asleep :stuck_out_tongue:

**Comment by firstred** [comment]
K. Whatever. I'm just going to close this as well, due to lack of response...
